### ISSUE TITLE
Parse.Query.subscribe() and Parse.Query.first() don't work in Parse v3.5.0

### ISSUE THREAD

**Issue by aarondill** [parse-community/Parse-SDK-JS#1596]
### New Issue Checklist

- [X] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [X] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [X] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).
- [X] I can reproduce the issue with the latest versions of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse JS SDK](https://github.com/parse-community/Parse-SDK-JS/releases). 

### Issue Description

Parse.Query.subscribe() and Parse.Query.first() throw errors upon calling
Query.first() is also recorded [here](https://github.com/parse-community/Parse-SDK-JS/issues/1592#issue-1432790510)
### Steps to reproduce

[Here is a Github Repository](https://github.com/aarondill/parseError) that includes files for use in testing this bug.
Create a Parse.Query and call subscribe() on it
alternatively create a LiveQueryClient and call subscribe() on it with a Parse.Query
### Actual Outcome

throws error "Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'socket')" upon calling.
### Expected Outcome

Expected a subscription object
###  Environment

Client
- Parse JS SDK version: [Parse JavaScript SDK v3.5.0](https://unpkg.com/parse@3.5.0/dist/parse.min.js)

**Comment by mtrezza** [comment]
What version of Parse Server are you using?

**Comment by mtrezza** [comment]
See https://github.com/parse-community/Parse-SDK-JS/issues/1593#issuecomment-1304288039

**Comment by aarondill** [comment]
> What version of Parse Server are you using?

I am using back4app. I'm not quite sure how to check the current server version but based on the other relevant reports, I would assume it's 4.5.0

**Comment by nekonenene** [comment]
I encountered the same problem.
https://github.com/nekonenene/back4app-test

```ts
const query = new Parse.Query('Person').addAscending('createdAt');
const firstPerson = await query.first();
```

This code makes the following error.

<img width="786" alt="image" src="https://user-images.githubusercontent.com/11713748/200160956-22e69310-1b52-41cc-bda9-7d535a3a4b2c.png">

<img width="875" alt="image" src="https://user-images.githubusercontent.com/11713748/200161000-e95e3114-5f7e-41ae-b377-e6df4b4d4de8.png">

I can avoid this error by writing:

```ts
const query = new Parse.Query('Person').addAscending('createdAt');
const [firstPerson] = await query.limit(1).find();
```

This problem occurs even after changing to v3.4.4 or v3.0.0, so I suspect that it is not a problem with Parse itself, may be caused by a dependent library or [back4app](https://www.back4app.com/) etc.

**Comment by mtrezza** [comment]
> This problem occurs even after changing to v3.4.4 or v3.0.0, so I suspect that it is not a problem with Parse itself, may be caused by a dependent library or [back4app](https://www.back4app.com/) etc.

What is the last version that works for you, where the problem is not occurring?

**Comment by aarondill** [comment]
> This problem occurs even after changing to v3.4.4 or v3.0.0, so I suspect that it is not a problem with Parse itself, may be caused by a dependent library or [back4app](https://www.back4app.com/) etc.

When I changed to [JS SDK v3.4.4](https://unpkg.com/parse@3.4.4/dist/parse.min.js), both first, and subscribe worked for me

**Comment by nekonenene** [comment]
Oh, really?
Then it seems that 3.5.0 keeps running in my environment instead of 3.4.4. Thank you.


> What is the last version that works for you, where the problem is not occurring?

I'm not sure because I haven't used Parse in the past. Sorry I can't help you.

**Comment by aarondill** [comment]
# Parse.Query.first() error reason and repair

## Relevant code from each version for query.first:

### [Parse JS JDK v3.4.4](https://unpkg.com/parse@3.4.4/dist/parse.js):

```js
{
    key: "first",
    value: function (options
    /*:: ?: FullOptions*/
    )
    /*: Promise<ParseObject | void>*/
    {
      var _this4 = this;

      options = options || {};
      var findOptions = {};

      if (options.hasOwnProperty('useMasterKey')) {
        findOptions.useMasterKey = options.useMasterKey;
      }

      if (options.hasOwnProperty('sessionToken')) {
        findOptions.sessionToken = options.sessionToken;
      }

      if (options.hasOwnProperty('context') && (0, _typeof2.default)(options.context) === 'object') {
        findOptions.context = options.context;
      }

      this._setRequestTask(findOptions);

      var controller = _CoreManager.default.getQueryController();

      var params = this.toJSON();
      params.limit = 1;
      var select = this._select;

      if (this._queriesLocalDatastore) {
        return this._handleOfflineQuery(params).then(function (objects) {
          if (!objects[0]) {
            return undefined;
          }

          return objects[0];
        });
      }

      return (0, _find.default)(controller).call(controller, this.className, params, findOptions).then(function (response) {
        var objects = response.results;

        if (!objects[0]) {
          return undefined;
        }

        if (!objects[0].className) {
          objects[0].className = _this4.className;
        } // Make sure the data object contains keys for all objects that
        // have been requested with a select, so that our cached state
        // updates correctly.


        if (select) {
          handleSelectResult(objects[0], select);
        }

        if (options.json) {
          return objects[0];
        } else {
          return _ParseObject.default.fromJSON(objects[0], !select);
        }
      });
    }
    /**
     *Comment describing eachBatch function here
     */
}
```

### [Parse JS JDK v3.5.0](https://unpkg.com/parse@3.5.0/dist/parse.js):

```js
{
    key: "first",
    value: function (options
    /*:: ?: FullOptions*/
    )
    /*: Promise<ParseObject | void>*/
    {
      options = options || {};
      var findOptions = {};

      if (options.hasOwnProperty('useMasterKey')) {
        findOptions.useMasterKey = options.useMasterKey;
      }

      if (options.hasOwnProperty('sessionToken')) {
        findOptions.sessionToken = options.sessionToken;
      }

      if (options.hasOwnProperty('context') && (0, _typeof2.default)(options.context) === 'object') {
        findOptions.context = options.context;
      }

      this._setRequestTask(findOptions);

      var controller = _CoreManager.default.getQueryController();

      var params = this.toJSON();
      params.limit = 1;
      var select = this._select;

      if (this._queriesLocalDatastore) {
        return this._handleOfflineQuery(params).then(function (objects) {
          if (!objects[0]) {
            return undefined;
          }

          return objects[0];
        });
      }

      return (0, _find.default)(controller).call(controller, this.className, params, findOptions).then(function (response) {
        var objects = response.results;

        if (!objects[0]) {
          return undefined;
        }

        if (!objects[0].className) {
          objects[0].className = this.className;
        } // Make sure the data object contains keys for all objects that
        // have been requested with a select, so that our cached state
        // updates correctly.


        if (select) {
          handleSelectResult(objects[0], select);
        }

        if (options.json) {
          return objects[0];
        } else {
          return _ParseObject.default.fromJSON(objects[0], !select);
        }
      });
    }
    /**
     *Comment describing eachBatch function here
    */
}
```

## Important _changes_

### v3.4.4:

Start of query.first function:

```js
var _this4 = this;
```

within return value.then():

```js
if (!objects[0].className) {
	objects[0].className = _this4.className;
}
```

### 3.5.0

start of query.first function removes \_this4 variable

within return value.then():

```js
if (!objects[0].className) {
	objects[0].className = this.className;
}
```

## Reason for error:

the return value of the first() function is:

```js
return (0, _find.default)(controller).call(controller, this.className, params, findOptions).then(function (response) {...})
```

It is important to note that the function within the then call will be run with _this_ set to undefined, due to [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode#no_this_substitution); while _\_this4_ in v3.4.4 is a Parse.Query object.

Due to this variation, while at first it seems to have the same result upon the removal of the \_this4 variable in favor of directly using the this value, due to the use of the variable in a seperate scope, the this value resolves to undefined and this is trying to access _undefined.className_ which (understandably) results in a typeError.

this change was likely made to save memory by the removal of a seemingly unnecessary variable, but due to the global scope of the then function, actually causes an error upon calling.

## Solution:

The readdition of the \_this4 variable to the top of the first function, and the modification of the assignment line to the code seen in v3.4.4 will resolve the error and return the correct ParseObject value.

the corrected code will be (identical to the v3.4.4 code):

```js
{
    key: "first",
    value: function (options
    /*:: ?: FullOptions*/
    )
    /*: Promise<ParseObject | void>*/
    {
      var _this4 = this;

      options = options || {};
      var findOptions = {};

      if (options.hasOwnProperty('useMasterKey')) {
        findOptions.useMasterKey = options.useMasterKey;
      }

      if (options.hasOwnProperty('sessionToken')) {
        findOptions.sessionToken = options.sessionToken;
      }

      if (options.hasOwnProperty('context') && (0, _typeof2.default)(options.context) === 'object') {
        findOptions.context = options.context;
      }

      this._setRequestTask(findOptions);

      var controller = _CoreManager.default.getQueryController();

      var params = this.toJSON();
      params.limit = 1;
      var select = this._select;

      if (this._queriesLocalDatastore) {
        return this._handleOfflineQuery(params).then(function (objects) {
          if (!objects[0]) {
            return undefined;
          }

          return objects[0];
        });
      }

      return (0, _find.default)(controller).call(controller, this.className, params, findOptions).then(function (response) {
        var objects = response.results;

        if (!objects[0]) {
          return undefined;
        }

        if (!objects[0].className) {
          objects[0].className = _this4.className;
        } // Make sure the data object contains keys for all objects that
        // have been requested with a select, so that our cached state
        // updates correctly.


        if (select) {
          handleSelectResult(objects[0], select);
        }

        if (options.json) {
          return objects[0];
        } else {
          return _ParseObject.default.fromJSON(objects[0], !select);
        }
      });
    }
    /**
     *Comment describing eachBatch function here
     */
}
```

I have so far only looked at the code differences for Parse.Query.first(), but I will soon look at Parse.Query.subscribe() and see if I can find the issue(s) that result in that error. My suspicion is that it is a similar or the same issue.

**Comment by aarondill** [comment]
# Parse.Query.subscribe() Error Reason and Solution

**Quick warning**, this comment is nearly identical to the comment above. To save reading, the v3.5.0 update removed a _\_this2_ variable that captures the _this_ context at the beginning of the subscribe function, which is then referenced within a then statement, whose this value is equal to undefined due to running in [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode#no_this_substitution).
## Relevant code from each version for query.first:

### [Parse JS JDK v3.4.4](https://unpkg.com/parse@3.4.4/dist/parse.js):

```js
{
    key: "subscribe",
    value: function (query
    /*: Object*/
    , sessionToken
    /*: ?string*/
    )
    /*: LiveQuerySubscription*/
    {
      var _this2 = this;

      if (!query) {
        return;
      }

      var className = query.className;
      var queryJSON = query.toJSON();
      var where = queryJSON.where;
      var fields = (0, _keys.default)(queryJSON) ? (0, _keys.default)(queryJSON).split(',') : undefined;
      var subscribeRequest = {
        op: OP_TYPES.SUBSCRIBE,
        requestId: this.requestId,
        query: {
          className: className,
          where: where,
          fields: fields
        }
      };

      if (sessionToken) {
        subscribeRequest.sessionToken = sessionToken;
      }

      var subscription = new _LiveQuerySubscription.default(this.requestId, query, sessionToken);
      this.subscriptions.set(this.requestId, subscription);
      this.requestId += 1;
      this.connectPromise.then(function () {
        _this2.socket.send((0, _stringify.default)(subscribeRequest));
      });
      return subscription;
    }
    /**
     * After calling unsubscribe you'll stop receiving events from the subscription object.
     *
     * @param {object} subscription - subscription you would like to unsubscribe from.
     */

  }
```

### [Parse JS JDK v3.5.0](https://unpkg.com/parse@3.5.0/dist/parse.js):

```js
{
    key: "subscribe",
    value: function (query
    /*: Object*/
    , sessionToken
    /*: ?string*/
    )
    /*: LiveQuerySubscription*/
    {
      if (!query) {
        return;
      }

      var className = query.className;
      var queryJSON = query.toJSON();
      var where = queryJSON.where;
      var fields = (0, _keys.default)(queryJSON) ? (0, _keys.default)(queryJSON).split(',') : undefined;
      var subscribeRequest = {
        op: OP_TYPES.SUBSCRIBE,
        requestId: this.requestId,
        query: {
          className: className,
          where: where,
          fields: fields
        }
      };

      if (sessionToken) {
        subscribeRequest.sessionToken = sessionToken;
      }

      var subscription = new _LiveQuerySubscription.default(this.requestId, query, sessionToken);
      this.subscriptions.set(this.requestId, subscription);
      this.requestId += 1;
      this.connectPromise.then(function () {
        this.socket.send((0, _stringify.default)(subscribeRequest));
      });
      return subscription;
    }
    /**
     * After calling unsubscribe you'll stop receiving events from the subscription object.
     *
     * @param {object} subscription - subscription you would like to unsubscribe from.
     */

  },{
    key: "subscribe",
    value: function (query
    /*: Object*/
    , sessionToken
    /*: ?string*/
    )
    /*: LiveQuerySubscription*/
    {
      if (!query) {
        return;
      }

      var className = query.className;
      var queryJSON = query.toJSON();
      var where = queryJSON.where;
      var fields = (0, _keys.default)(queryJSON) ? (0, _keys.default)(queryJSON).split(',') : undefined;
      var subscribeRequest = {
        op: OP_TYPES.SUBSCRIBE,
        requestId: this.requestId,
        query: {
          className: className,
          where: where,
          fields: fields
        }
      };

      if (sessionToken) {
        subscribeRequest.sessionToken = sessionToken;
      }

      var subscription = new _LiveQuerySubscription.default(this.requestId, query, sessionToken);
      this.subscriptions.set(this.requestId, subscription);
      this.requestId += 1;
      this.connectPromise.then(function () {
        this.socket.send((0, _stringify.default)(subscribeRequest));
      });
      return subscription;
    }
    /**
     * After calling unsubscribe you'll stop receiving events from the subscription object.
     *
     * @param {object} subscription - subscription you would like to unsubscribe from.
     */
  }
```

## Important _changes_

### v3.4.4:

Start of query.subscribe function:

```js
var _this2 = this;
```

just before return(error thrown in v4.5.0)

```js
this.connectPromise.then(function () {
	_this2.socket.send((0, _stringify.default)(subscribeRequest));
});
```

### v3.5.0

start of Query.subscribe function removes \_this2 variable

just before return(error thrown in v4.5.0):

```js
this.connectPromise.then(function () {
	this.socket.send((0, _stringify.default)(subscribeRequest));
});
```

## Reason for error:

the v3.5.0 code uses this line:

```js
this.connectPromise.then(function () {
	this.socket.send((0, _stringify.default)(subscribeRequest));
});
```

Similarly to the Parse.Query.first() issue, it is important to note that the function within the then call will be run with _this_ set to undefined due to [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode#no_this_substitution); while _\_this2_ in v3.4.4 is a Parse.Query object.

Due to this variation, while at first it seems to have the same result upon the removal of the \_this2 variable in favor of directly using the this value, due to the use of the variable in a seperate scope, the this value resolves to undefined and this is trying to access _undefined.socket_ which (understandably) results in a typeError.

this change was likely made to save memory by the removal of a seemingly unnecessary variable, but due to the global scope of the then function, actually causes an error upon calling.

## Solution:

The readdition of the \_this2 variable to the top of the subscribe function, and the modification of the code to the code seen in v3.4.4 will resolve the error and return the correct LiveQuerySubscription value.

the corrected code will be (identical to the v3.4.4 code):

```js
{
    key: "subscribe",
    value: function (query
    /*: Object*/
    , sessionToken
    /*: ?string*/
    )
    /*: LiveQuerySubscription*/
    {
      var _this2 = this;

      if (!query) {
        return;
      }

      var className = query.className;
      var queryJSON = query.toJSON();
      var where = queryJSON.where;
      var fields = (0, _keys.default)(queryJSON) ? (0, _keys.default)(queryJSON).split(',') : undefined;
      var subscribeRequest = {
        op: OP_TYPES.SUBSCRIBE,
        requestId: this.requestId,
        query: {
          className: className,
          where: where,
          fields: fields
        }
      };

      if (sessionToken) {
        subscribeRequest.sessionToken = sessionToken;
      }

      var subscription = new _LiveQuerySubscription.default(this.requestId, query, sessionToken);
      this.subscriptions.set(this.requestId, subscription);
      this.requestId += 1;
      this.connectPromise.then(function () {
        _this2.socket.send((0, _stringify.default)(subscribeRequest));
      });
      return subscription;
    }
    /**
     * After calling unsubscribe you'll stop receiving events from the subscription object.
     *
     * @param {object} subscription - subscription you would like to unsubscribe from.
     */

  }
```

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [3.5.1-alpha.2](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.1-alpha.2)

**Comment by mtrezza** [comment]
Closing via https://github.com/parse-community/Parse-SDK-JS/pull/1600

**Comment by mtrezza** [comment]
This should be fixed with https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.1-alpha.2, could you try it out and let us know?

**Comment by ChrisGrant** [comment]
> This should be fixed with https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.1-alpha.2, could you try it out and let us know?

It fixed the problem for me. I was on 3.5.0 and was seeing the same issue. Upgrading to 3.5.1-alpha2 makes the problem go away.

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [3.5.1-beta.2](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.1-beta.2)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [3.5.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.1)
